myst_nb/converter.py

```diff
@@ -0,0 +1,239 @@
+        .get("text_representation", {})
+        .get("format_name", None)
+        != "myst"
+    ):
+        return False
+
+    if "name" not in front_matter.get("kernelspec", {}):
+        raise IOError(
+            "A myst notebook text-representation requires " "kernelspec/name metadata"
+        )
+    if "display_name" not in front_matter.get("kernelspec", {}):
+        raise IOError(
+            "A myst notebook text-representation requires "
+            "kernelspec/display_name metadata"
+        )
+    return True
+
+
+class MystMetadataParsingError(Exception):
+    """Error when parsing metadata from myst formatted text"""
+
+
+def strip_blank_lines(text):
+    text = text.rstrip()
+    while text and text.startswith("\n"):
+        text = text[1:]
+    return text
+
+
+class MockDirective:
+    option_spec = {"options": True}
+    required_arguments = 0
+    optional_arguments = 1
+    has_content = True
+
+
+def read_fenced_cell(token, cell_index, cell_type):
+    from myst_parser.parse_directives import DirectiveParsingError, parse_directive_text
+
+    try:
+        _, options, body_lines = parse_directive_text(
+            directive_class=MockDirective,
+            argument_str="",
+            content=token.content,
+            validate_options=False,
+        )
+    except DirectiveParsingError as err:
+        raise MystMetadataParsingError(
+            "{0} cell {1} at line {2} could not be read: {3}".format(
+                cell_type, cell_index, token.map[0] + 1, err
+            )
+        )
+    return options, body_lines
+
+
+def read_cell_metadata(token, cell_index):
+    metadata = {}
+    if token.content:
+        try:
+            metadata = json.loads(token.content.strip())
+        except Exception as err:
+            raise MystMetadataParsingError(
+                "Markdown cell {0} at line {1} could not be read: {2}".format(
+                    cell_index, token.map[0] + 1, err
+                )
+            )
+        if not isinstance(metadata, dict):
+            raise MystMetadataParsingError(
+                "Markdown cell {0} at line {1} is not a dict".format(
+                    cell_index, token.map[0] + 1
+                )
+            )
+
+    return metadata
+
+
+def myst_to_notebook(
+    text,
+    code_directive=CODE_DIRECTIVE,
+    raw_directive=RAW_DIRECTIVE,
+    add_source_map=False,
+):
+    """Convert text written in the myst format to a notebook.
+
+    :param text: the file text
+    :param code_directive: the name of the directive to search for containing code cells
+    :param raw_directive: the name of the directive to search for containing raw cells
+    :param add_source_map: add a `source_map` key to the notebook metadata,
+        which is a list of the starting source line number for each cell.
+
+    :raises MystMetadataParsingError if the metadata block is not valid JSON/YAML
+
+    NOTE: we assume here that all of these directives are at the top-level,
```

Comment:
This is a pretty big assumption, no? So myst-nb will not support nested directives?

Reply:
Not all directives, just `{code-cell}`. This is necessary for compliance with round trip conversion with notebooks, as obviously you can't have a code cell nested in a markdown cell. When I say 'assume`, I mean we just ignore any nested code cells, which will, if present, then raise an error on the sphinx parse, because it won't have a directive associated with it.

Reply:
Ah - that makes sense then, since `{code-cell}` doesn't support myst-markdown

Reply:
Yep thats the one 'compromise' we make, in terms of treating the whole file as a notebook (rather than 'extracting' notebooks from files, like jupyter-sphinx does): you can't do something like

`````
````{note}
```{code-cell}
a = 1
```
````
`````

with the benefit of being abled to treat them exactly the same a standard notebooks; integrating them with jupyter-cache etc.